src/Stratis.Bitcoin.Features.BlockStore/ProvenHeadersBlockStoreSignaled.cs

```diff
@@ -99,9 +100,20 @@ private void CreateAndStoreProvenHeader(int blockHeight, ChainedHeaderBlock chai
             uint256 provenHeaderHash = newProvenHeader.GetHash();
             this.provenBlockHeaderStore.AddToPendingBatch(newProvenHeader, new HashHeightPair(provenHeaderHash, blockHeight));
 
-            logger.LogTrace("Created Proven Header at height {0} with hash {1} and adding to the pending batch to be stored.", blockHeight, provenHeaderHash);
+            this.logger.LogTrace("Created Proven Header at height {0} with hash {1} and adding to the pending batch to be stored.", blockHeight, provenHeaderHash);
 
-            chainedHeaderBlock.SetHeader(newProvenHeader);
+            // If our node is in IBD the block will not be announced to peers.
+            // If not in IBD the signaler may expect the block header to be of type PH.
+            // TODO: Memory foot print:
+            // This design will cause memory to grow over time (depending on how long the node is running)
+            // based on the size of the Proven Headers (a proven header can be up to 1000 bytes).
+            // This is also correct for regular header (which are 80 bytes in size).
+            // If we want to be able to control the size of PH we will need to change the logic
+            // in ProvenHeadersBlockStoreBehavior and load the PH from the PH store instead
+            if (!this.initialBlockDownloadState.IsInitialBlockDownload())
```

Comment:
I think there is a race condition. 

We are in BlockStoreSignaled.OnNextCore and we have a new block. We call AddBlockToQueue which leads to ProvenHeadersBlockStoreSignaled.AddBlockToQueue. Since we are in IBD we don't set PH. Now we proceed to BlockStoreSignaled line 90 and by that time we are not in IBD anymore so we add header to announce queue. 
Then we announce it using ProvenHeadersBlockStoreBehavior.BuildAnnouncedHeaderPayload where we fail at 72 because it's not a PH.

Reply:
Yes it is a race good spot, suggestions to fix? (apart for loading PH in the the behaviour)

Reply:
Cache result of IsInitialBlockDownload and use it everywhere in this flow so it can't be changed due to the race condition.

Reply:
fixed